docs/src/rfcs/00X-global-buffer.md

```diff
@@ -0,0 +1,324 @@
+
+    pub(crate) trait GetExt<I> {
+        type Output;
+        fn get(&self, index: I) -> Option<Self::Output>;
+        unsafe fn get_unchecked(&self, index: I) -> Self::Output;
+    }
+
+    pub(crate) trait GetMutExt<I> {
+        type Output;
+        fn get_mut(&mut self, index: I) -> Option<Self::Output>;
+        unsafe fn get_unchecked_mut(&mut self, index: I) -> Self::Output;
+    }
+
+
+    impl<'a, T> GetExt<GlobalIndex> for Global<'a, [T]> {
+        type Output = Global<'a, T>;
+        fn get(&self, index: GlobalIndex) -> Option<Self::Output> {
+            self.get(index.0)
+        }
+        unsafe fn get_unchecked(&self, index: GlobalIndex) -> Self::Output [
+            self.get_unchecked(index.0)
+        }
+    }
+
+    impl<'a, T> GetExt<GlobalIndex> for GlobalMut<'a, [T]> {
+        type Output = Global<'a, T>;
+        fn get(&self, index: GlobalIndex) -> Option<Self::Output> {
+            self.get(index.0)
+        }
+        unsafe fn get_unchecked(&self, index: GlobalIndex) -> Self::Output [
+            self.get_unchecked(index.0)
+        }
+    }
+
+    impl<'a, T> GetMutExt<GlobalIndex> for GlobalMut<'a, [T]> {
+        type Output = GlobalMut<'a, T>;
+        fn get_mut(&mut self, index: GlobalIndex) -> Option<Self::Output> {
+            self.get_mut(index.0)
+        }
+        unsafe fn get_unchecked_mut(&mut self, index: GlobalIndex) -> Self::Output [
+            self.get_unchecked_mut(index.0)
+        }
+    }
+    
+## Example 
+
+As an example, suppose that a hypothetical "zip_mut_with" (see ndarray [zip_mut_with](https://docs.rs/ndarray/0.13.1/ndarray/struct.ArrayBase.html#method.zip_mut_with)) function was added to spirv-std. 
+
+    // spirv-std/src/lib.rs
+    
+    impl<'a, T, const N: usize> GlobalBufferMut<[T; N]> {
+        pub fn zip_mut_with<C: Copy>(mut self, rhs: &GlobalBuffer<[T; N]>, constants: C, f: impl fn(GloablMut<T>, Global<T>, C)) {
+            let index = global_index();
+            barrier(); // barrier for any previous writes to self
+            unsafe {
+                self.as_unsafe_mut_slice().get_mut(index).zip(rhs.as_slice().get(index))
+                    .map(|(lhs, rhs)| f(lhs, rhs, constants));
+            }
+            // this method consumes self
+            // alternatively take self by reference, and emit a barrier here
+        }
+    }
+  
+    // scaled_add.rs
+    use spirv_std::{Buffer, BufferMut};
+    
+    type T = f32;
+    const N: usize = 1024;
+    
+    #[allow(unused_attributes)]
+    #[spirv(gl_compute(local_size=64)]
+    pub fn scaled_add(
+        #[spirv(descriptor_set=1, binding=0)] x: GlobalBuffer<[T; N]>,
+        #[spirv(descriptor_set=1, binding=1)] mut y: GlobalBufferMut<[T; N]>,
```

Comment:
I have some slight trouble understanding this as written - does this mean that each thread gets 1024 items available to itself to iterate over, or does this mean that `N` here is the amount of total threads launched? In the latter case, shouldn't this be `GlobalBuffer<[T]>` instead from a usability point of view? I see in your proposal that this is used to create a slice of N long; however this would prose problems for most use-cases as they won't know the size of the `GlobalBuffer` at compile-time right?

It's kind of unfortunate that we can't rely on vulkan's built in [robust buffer access](https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#features-robustBufferAccess) support since it just silently ignores out of bounds access

Reply:
The intent is to have both arrays and runtime arrays, with GlobalBuffer<[T; N]> and GlobalBuffer<[T]>. Right now, slices can't be used in entry parameters. 
In this simple example, the buffers have 1024 f32's. I haven't proposed any way to validate that the buffers actually have that length. You are right that generally these would be runtime arrays with some runtime length. 
The compute shader sees 1024 items in each buffer. It is up to GlobalBufferMut, through some safe interface, to ensure that each thread / invocation only sees some exclusive piece of it. In this case, the zip_mut_with function just indexes both buffers by the global index and applies the provided function, which operates on a single pair of scalars. If the global index is out of bounds, the function is not applied.
I think I will edit this to show using slices because the array is confusing.